torch/optim/lr_scheduler.py

```diff
@@ -427,3 +427,85 @@ def state_dict(self):
     def load_state_dict(self, state_dict):
         self.__dict__.update(state_dict)
         self._init_is_better(mode=self.mode, threshold=self.threshold, threshold_mode=self.threshold_mode)
+
+
+class CosineAnnealingWarmRestarts(_LRScheduler):
+    r"""Set the learning rate of each parameter group using a cosine annealing
+    schedule, where :math:`\eta_{max}` is set to the initial lr, :math:`T_{cur}`
+    is the number of epochs since the last restart and :math:`T_{i}` is the number
+    of epochs between two warm restarts in SGDR:
+
+    .. math::
+
+        \eta_t = \eta_{min} + \frac{1}{2}(\eta_{max} - \eta_{min})(1 +
+        \cos(\frac{T_{cur}}{T_{i}}\pi))
+
+    When :math:`\T_{cur}=T_{i}`, set :math:`\eta_t = \eta_{min}`.
+    When :math:`\T_{cur}=0`(after restart), set :math:`\eta_t=\eta_{max}`.
+
+    It has been proposed in
+    `SGDR: Stochastic Gradient Descent with Warm Restarts`_.
+
+    Args:
+        optimizer (Optimizer): Wrapped optimizer.
+        T_0 (int): Number of iterations for the first restart.
+        T_mult (int, optional): A factor increases :math:`\T_{i}` after a restart.
+            If T_mult is float, T_mult will be converted to int. Default: 1.
```

Comment:
Shall we do an explicit type check, and raise an error if it is not an int? This also applies to `T_0`. We might want to avoid silently surprising people even if they did pay attention to the doc.

Reply:
Ok, thank you for the suggestions.